File: backend/app/core/policy.py
"""Confirmation policy is rules-based. The planner cannot lower required confirmation."""

from __future__ import annotations

from enum import Enum

from app.ai.models import ActionType, Confirmation


class PolicyFloor(str, Enum):
    AUTO = "auto"
    CONFIRM = "confirm"
    HANDOFF = "handoff"


DEFAULT_FLOORS: dict[ActionType, Confirmation] = {
    ActionType.CREATE_REMINDER: Confirmation.AUTO,
    ActionType.CREATE_TASK: Confirmation.AUTO,
    ActionType.COMPLETE_TASK: Confirmation.AUTO,
    ActionType.CREATE_CALENDAR_EVENT: Confirmation.AUTO,
    ActionType.UPDATE_CALENDAR_EVENT: Confirmation.CONFIRM,
    ActionType.DELETE_CALENDAR_EVENT: Confirmation.CONFIRM,
    ActionType.QUERY_CALENDAR: Confirmation.AUTO,
    ActionType.DRAFT_EMAIL: Confirmation.AUTO,
    ActionType.SEND_EMAIL: Confirmation.CONFIRM,
    ActionType.SEARCH_EMAIL: Confirmation.AUTO,
    ActionType.CALL_CONTACT: Confirmation.CONFIRM,
    ActionType.PREPARE_SMS: Confirmation.HANDOFF,
    ActionType.PREPARE_WHATSAPP: Confirmation.HANDOFF,
    ActionType.PREPARE_TELEGRAM: Confirmation.HANDOFF,
    ActionType.SHARE_MEDIA: Confirmation.CONFIRM,
    ActionType.SAVE_NOTE: Confirmation.AUTO,
    ActionType.QUERY_TASKS: Confirmation.AUTO,
    ActionType.QUERY_MEMORY: Confirmation.AUTO,
}

_RANK = {
    Confirmation.AUTO: 0,
    Confirmation.CONFIRM: 1,
    Confirmation.HANDOFF: 2,
}


def apply_policy_floor(action_type: ActionType, proposed: Confirmation) -> Confirmation:
    """Raise confirmation to at least the floor. Never reduce it."""
    floor = DEFAULT_FLOORS[action_type]
    if _RANK[proposed] >= _RANK[floor]:
        return proposed
    return floor

File: mobile/lib/core/actions/policy.dart
import 'models.dart';

const _rank = {'auto': 0, 'confirm': 1, 'handoff': 2};

const defaultFloors = {
  'CREATE_REMINDER': 'auto',
  'CREATE_TASK': 'auto',
  'COMPLETE_TASK': 'auto',
  'CREATE_CALENDAR_EVENT': 'auto',
  'UPDATE_CALENDAR_EVENT': 'confirm',
  'DELETE_CALENDAR_EVENT': 'confirm',
  'QUERY_CALENDAR': 'auto',
  'DRAFT_EMAIL': 'auto',
  'SEND_EMAIL': 'confirm',
  'SEARCH_EMAIL': 'auto',
  'CALL_CONTACT': 'confirm',
  'PREPARE_SMS': 'handoff',
  'PREPARE_WHATSAPP': 'handoff',
  'PREPARE_TELEGRAM': 'handoff',
  'SHARE_MEDIA': 'confirm',
  'SAVE_NOTE': 'auto',
  'QUERY_TASKS': 'auto',
  'QUERY_MEMORY': 'auto',
};

/// Rules engine. Planner output cannot lower confirmation below the floor.
String applyPolicyFloor(String actionType, String proposed) {
  final floor = defaultFloors[actionType] ?? 'confirm';
  final proposedRank = _rank[proposed] ?? 0;
  final floorRank = _rank[floor] ?? 1;
  return proposedRank >= floorRank ? proposed : floor;
}

bool isMessagingHandoff(String actionType) {
  return actionType == 'PREPARE_WHATSAPP' ||
      actionType == 'PREPARE_SMS' ||
      actionType == 'PREPARE_TELEGRAM';
}

/// WhatsApp stays a handoff (not marked sent). If the person is already in
/// memory or the user just picked them, skip the extra "Go ahead" tap.
bool needsUserGoAhead(PlannedAction action, {bool identityConfirmed = false}) {
  final policy = applyPolicyFloor(action.type, action.confirmation);
  if (policy == 'auto') return false;
  if (identityConfirmed && isMessagingHandoff(action.type)) return false;
  return true;
}
